fontforge: support styleMap attribute as a subset of OS/2 fsSelection

adrientetar opened the pull request:
Closes #2131.

r? @frank-trampe

frank-trampe commented:
@adrientetar, the Travis build failed.

```
splineutil.c: In function ‘SplineFontFree’:
splineutil.c:6470:67: error: ‘styleMapFamilyName’ undeclared (first use in this function)
splineutil.c:6470:67: note: each undeclared identifier is reported only once for each function it appears in
```

adrientetar commented:
Updated.

frank-trampe commented:
@adrientetar, can you explain precisely what you're doing in this pull request?

adrientetar commented:
I added a styleMap attribute which in turn is a subset of OS/2 fsStyle in order to effectively support styleMapStyleName with possibility for the user to change the setting before export (added a widget for this). I used a bitmask as the repr which corresponds to the OS/2 attribute. This is also used in TTFs if set instead of copying the macStyle attribute (which is corrected to be inline with the styleMap if needed).  
Besides that I implemented fallback mechanisms for various attributes. The styleMapFamilyName fallback is made the same as that of robofab and ufo2fdk.  
I removed OS2FamilyName and OS2StyleName because these were not really OS/2 attributes anyway. And I added support for reading the styleMap out of a TTF.  
I think that's pretty much it. The goal is to standardize styleMapFamilyName, and to support styleMapStyleName properly and with a fallback so you don't need to set it if you use default style mapping attributes. This is important for robofont or ufo2fdk and permits style-linking on Windows for the generated files. Style linking used for the "Bold" or "Italic" buttons or MS Word for instance and style linking is important for grouping fonts properly in certain Windows applications.

adrientetar commented:
Say, with the current code I have no way to set up `styleMapStyleName` unless picked from a foreign UFO. The current code guarantees round-tripping but it doesn't have those attributes supported "natively".

adrientetar commented:
Bump.

davelab6 commented:
Frank is mostly offline this week :/

frank-trampe commented:
I'm looking at this, though.

adrientetar commented:
No problem, I was just worried that this had been forgotten.

frank-trampe commented on fontforge/ufo.c:
```diff
@@ -1086,8 +1103,34 @@ static int UFOOutputFontInfo(const char *basedir, SplineFont *sf, int layer) {
     PListAddString(dictnode,"familyName",sf->familyname_with_timestamp ? sf->familyname_with_timestamp : sf->familyname);
     PListAddString(dictnode,"styleName",SFGetModifiers(sf));
-    if (sf->pfminfo.os2_family_name != NULL) PListAddString(dictnode,"styleMapFamilyName", sf->pfminfo.os2_family_name);
-    if (sf->pfminfo.os2_style_name != NULL) PListAddString(dictnode,"styleMapStyleName", sf->pfminfo.os2_style_name);
```
I don't think that we want to add these values automatically if we are exporting something that we imported from U. F. O. without those values.

adrientetar replied:
Good point, I'll change that.

frank-trampe commented on fontforgeexe/fontinfo.c:
```diff
@@ -196,6 +196,15 @@ static GTextInfo ibmfamily[] = {
     { (unichar_t *) N_("Sy Miscellaneous"), NULL, 0, 0, (void *) 0xc0f, NULL, 0, 0, 0, 0, 0, 0, 1, 0, 0, '\0'},
     GTEXTINFO_EMPTY
 };
+static GTextInfo stylemap[] = {
```
I find the casting sort of confusing. How about putting an annotated set or array of integers before this and storing pointers into it? (I'm not saying that's the best approach.) There at least needs to be a comment describing what is happening.

adrientetar replied:
Well I just copied the structure of the surrounding elements.

frank-trampe replied:
Oh. You're right. A comment then.

adrientetar replied:
I didn't make this repr so I have nothing particular to say about it. If you want me write something before the stack of arrays, I can do that.

frank-trampe replied:
```
# Note that we are storing integer data in a pointer field, as others seem to have done before us, hence the casts. They are not to be dereferenced.
```

frank-trampe commented on fontforgeexe/fontinfo.c:
```diff
@@ -8433,16 +8447,6 @@ return;
     vgcd[13].gd.u.list = ibmfamily;
     vgcd[13].creator = GListButtonCreate;
 
```
Why is this getting dropped?

frank-trampe replied:
Or moved, rather?

adrientetar replied:
Because the concerned checkbox is at the end of the window and its positioning is relative to that of the element I added (statically so).

frank-trampe replied:
Does this break the tab order?

adrientetar replied:
I don't understand, what do you mean?

frank-trampe replied:
If you tab through the elements in the form, does the cursor move as expected?

adrientetar replied:
I tried to Tab but it really only works with numeric fields, I see no outline and cannot open up list elements when tabbing. 
I'll note that there is existing disorder, for instance OS/2 Version which is the first field of this tab is ordered towards the bottom. Anyways, tab navigation isn't useable (at least on my computer) for this kind of elements.

frank-trampe replied:
Tabbing does not work for me, either, as it happens, and you are right that the order is already broken anyway. So it's probably not worth fixing in this pull request.

frank-trampe commented on fontforgeexe/math.c:
```diff
@@ -1799,6 +1799,7 @@ static void MKDInit(MathKernDlg *mkd,SplineChar *sc) {
     mkd->dummy_sf.glyphs = mkd->chars;
     mkd->dummy_sf.glyphcnt = mkd->dummy_sf.glyphmax = 4;
     mkd->dummy_sf.pfminfo.fstype = -1;
+    mkd->dummy_sf.pfminfo.stylemap = -1;
```
Do we perhaps need a function that takes a pointer to a dummy_sf and initializes it as a dummy?

frank-trampe replied:
(Not that that's a defect in this pull request.)

adrientetar replied:
I don't esp. want to tackle this just now, but you are right that there is duplication with the other function.

adrientetar commented:
Updated.

frank-trampe commented:
Oops. I missed something.

frank-trampe commented on fontforge/sfd.c:
```diff
@@ -7465,14 +7465,6 @@ bool SFD_GetFontMetaData( FILE *sfd,
-    else if ( strmatch(tok,"OS2StyleName:")==0 )
-    {
-	if (!sf->pfminfo.os2_style_name) sf->pfminfo.os2_style_name = SFDReadUTF7Str(sfd);
-    }
```
In the interest of maximizing backwards compatibility, we ought to process these directives into the new data types with a note that they are deprecated.

frank-trampe commented on fontforge/sfd.c:
```diff
@@ -7895,6 +7891,10 @@ bool SFD_GetFontMetaData( FILE *sfd,
     }
+    else if ( strmatch(tok,"StyleMap:")==0 )
+    {
+    gethex(sfd,(uint32 *)&sf->sfntRevision);
```
sfntRevision?

adrientetar replied:
I think I noticed it simultanouesly to you while looking at your previous comment, haha. Updated.

frank-trampe commented:
Oops! I gave you a shell script comment instead of a C comment.

frank-trampe commented:
Also see my note about the backward compatibility in the s. f. d. parser.

adrientetar commented:
Updated. I have added backwards-compat reading of SFD attributes.

frank-trampe commented on fontforge/ufo.c:
```diff
@@ -1086,8 +1103,38 @@ static int UFOOutputFontInfo(const char *basedir, SplineFont *sf, int layer) {
+        char* styleMapFamily;
+        if (sf->styleMapFamilyName != NULL) {
+            /* Empty styleMapStyleName means we imported a UFO that does not have this field. Bypass the fallback. */
+            if (sf->styleMapFamilyName[0]!='\0')
```
I would point out that this would fail to round-trip a blank name, but I think that a name cannot be blank.

adrientetar replied:
Yea I think editors don't output an empty field.

frank-trampe commented:
I'm fine with this.
